### ISSUE TITLE
Support for richer nested operation during create/delete

### ISSUE THREAD

**Issue by fchu** [prisma/prisma#3538]
## Problem

A bunch of nested write operations during updates are provided, namely:

* ```create```
* ```connect```
* ```connectOrCreate```
* ```update```
* ```upsert```
* ```delete```

However, create operations only offer a subset:

* ```create```
* ```connect```
* ```connectOrCreate```

while delete operations don't allow nested writes at all.

This is unfortunate because there are times where after an update + connect or a delete, you want to transactionally update a relation as well (simple example: a reference counter of sorts in a different table)

## Suggested solution

At a high-level, support mixtures of `connect`/`update` as nested operations for `update`/`delete` top-level operations, eg maybe a `connectAndUpdate` within update, or a `updateBeforeDelete` within delete, though I don't quite like the idea of creating more action verbs. Ideally it would be some expressive semantics so it can be combined in various situation, eg:
```ts
prima.user.create({
  data: {
    name,
    group: [
      { connect: { id: groupId } },
      { update: { moneyRaised: { increment: contributionAmount } } },
    ],
  },
});
```

I recognize there are challenges, eg:
- what if a nested update in a delete is on a relation that has a DB-level `on cascade delete`?
- in what order should updates and connects happen?
But I feel those could be solvable? (cascade representation in `schema.prisma`, support for operation ordering)

## Alternatives

I suppose #1844 could solve this need adequately, but also seems to be an over-kill solution, especially since nested operations is already an existing paradigm with existing semantics.

## Additional context
n/a

**Comment by matthewmueller** [comment]
Thanks @fchu for raising this. I think I understand the motivation at a high-level, we offer some operations like `connectOrCreate`, but not other operations like `deleteAndDisconnect`.

It'd be helpful for us to better understand what queries can't be expressed at all or what queries can't be expressed easily with the current API. Isolated examples would be most helpful.

I suspect that the transaction API will cover most of these cases: https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/transactions#bulk-operations-experimental. But there's certainly an opportunity to add more operations in places where this breaks down.

**Comment by matthewmueller** [comment]
Yep, queries that depend on previous query results are tricky. We currently suggest pre-computing the IDs using UUID or similar. We're also thinking about ways to improve this though.

**Comment by jamesguan** [comment]
> Yep, creating transactions where the query that depend on previous query results is currently tricky. We currently suggest pre-computing the IDs using UUID or similar. We're also thinking about ways to improve this though.

This is a pretty great idea. This would only work on one level of creation though.

**Comment by janpio** [comment]
> Yep, creating transactions where the query that depend on previous query results is currently tricky.

This has changed with interactive transactions: https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions There you have transaction guarantees, but can use data from the first query in the second one.

**Comment by fchu** [comment]
Hi Jan, thanks for the follow up. This sounds great and though not as compact as a nested operation, is probably a more flexible and powerful solution overall. Thanks for implementing it, feel free to close this issue as you see fit!

**Comment by luxaritas** [comment]
I'd still like to see extended nested operations - it would help keep queries more readable in some situations

**Comment by 0xdevalias** [comment]
> This has changed with interactive transactions: [prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions) There you have transaction guarantees, but can use data from the first query in the second one.

While this is definitely a solution now, I agree with the following:

> I'd still like to see extended nested operations - it would help keep queries more readable in some situations

Currently it feels confusing that I can use nested writes/etc in some places, but not others, and it complicates the DX of using prisma. I think it would be helpful if it was standardised one way or the other (with my preference being to add support for missing 'nested write' aspects)

See also:

- #6440 

> It'd be helpful to better understand what queries can't be expressed at all or what queries can't be expressed easily with the current API. Targeted, isolated examples would be most useful for us.

My specific use case is creating a new `Response`, and then updating a `status` field on it's linked `Inquiry`; which currently I would achieve within an interactive (or I guess even sequential) transaction:

```ts
ctx.db.$transaction(async (dbTxn) => {
  dbTxn.response.create({
    data: {
      inquiryId: input.inquiryId,
      content: input.content,
      createdById: ctx.session.user.id,
    },
  });

  dbTxn.inquiry.update({
    where: {
      id: input.inquiryId,
      status: "new",
    },
    data: {
      status: "in_progress",
    },
  });
}
```

But would be nice if I could do something like the following:

- https://www.prisma.io/docs/orm/reference/prisma-client-reference#update-1

```ts
// Create a new response
return ctx.db.response.create({
  data: {
    inquiryId: input.inquiryId,
    content: input.content,
    createdById: ctx.session.user.id,
    inquiry: { update: { status: "in_progress" } },
  },
});
```

Or perhaps even:

```ts
// Create a new response
return ctx.db.response.create({
  data: {
    inquiryId: input.inquiryId,
    content: input.content,
    createdById: ctx.session.user.id,
    inquiry: {
      update: {
        status: "in_progress",
        where: { status: "new" },
      },
    },
  },
});
```
